### PR TITLE
Add reshape operator

### LINKED ISSUE #4009: Reshape Operator.

### PR DESCRIPTION
**Pull request by kuke**
Resolve #4009

### CODE REVIEW

File: paddle/operators/reshape_op.cc [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  ReshapeOp(const std::string &type, const framework::VariableNameMap &inputs,
+            const framework::VariableNameMap &outputs,
+            const framework::AttributeMap &attrs)
+      : OperatorWithKernel(type, inputs, outputs, attrs) {}
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    auto *in = ctx.Input<framework::Tensor>("X");
+    auto shape = ctx.Attr<std::vector<int>>("shape");
+    PADDLE_ENFORCE_EQ((unsigned)shape.size(), in->dims().size(),
```
Comment (dzhwinter) [inline paddle/operators/reshape_op.cc]: `unsigned long` please. the size returned by ddim is ssize_t, which is 8byte width.

File: paddle/operators/reshape_op.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+#pragma once
+
+#include "paddle/framework/eigen.h"
+#include "paddle/framework/op_registry.h"
+
+namespace paddle {
+namespace operators {
+
+using Tensor = framework::Tensor;
```
Comment (dzhwinter) [inline paddle/operators/reshape_op.h]: do not use `using` in the header file, according to the google style.

File: python/paddle/v2/framework/tests/test_reshape_op.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+    def setUp(self):
+        self.type = "reshape"
+        self.inputs = {'X': np.random.random((37, 51)).astype("float32"), }
+        self.attrs = {'shape': [51, 37]}
+        self.outputs = {'Out': self.inputs['X'].reshape(self.attrs['shape'])}
+
+
+class ReshapeGradOpTest(GradientChecker):
+    def test_normal(self):
```
Comment (dzhwinter) [inline python/paddle/v2/framework/tests/test_reshape_op.py]: now can use the function `setUp` to generate forward operator.
please take a look at [this case](https://github.com/PaddlePaddle/Paddle/pull/3888/files#diff-c2dd71df56cae469f7474f18bea1d8a0R70)

File: paddle/operators/reshape_op.h [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+template <typename Place, typename T>
+class ReshapeKernel : public framework::OpKernel {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const {
+    auto* out = ctx.Output<Tensor>("Out");
+    auto* in = ctx.Input<Tensor>("X");
+    out->mutable_data<T>(ctx.GetPlace());
+
+    auto shape = ctx.Attr<std::vector<int>>("shape");
+    std::vector<int64_t> tmp;
```
Comment (dzhwinter) [inline paddle/operators/reshape_op.h]: Please do not mix use of `int64_t` and `size_t`, which is ugly for users who read our code.

File: paddle/operators/reshape_op.cc (hunk 1 above)
Comment (kuke) [reply paddle/operators/reshape_op.cc]: Removed this unreasonable line.

File: paddle/operators/reshape_op.h (hunk 2 above)
Comment (kuke) [reply paddle/operators/reshape_op.h]: Done

File: paddle/operators/reshape_op.h (hunk 4 above)
Comment (kuke) [reply paddle/operators/reshape_op.h]: Done

File: python/paddle/v2/framework/tests/test_reshape_op.py (hunk 3 above)
Comment (kuke) [reply python/paddle/v2/framework/tests/test_reshape_op.py]: Done

File: paddle/operators/reshape_op.cc [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+      if (capacity < 0) {
+        capacity = dim;
+      } else {
+        capacity *= dim;
+      }
+    }
+    int64_t in_size = framework::product(in->dims());
+    PADDLE_ENFORCE_EQ(capacity, in_size,
+                      "The size of Input(X) mismatches with Attr(shape).");
+    ctx.Output<framework::Tensor>("Out")->Resize(in->dims());
```
Comment (qingqing01) [inline paddle/operators/reshape_op.cc]: Why not  `Resize(shape)`?  The dims of output is not `shape` ?

File: paddle/operators/reshape_op.cc [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+  ReshapeOpMaker(framework::OpProto *proto,
+                 framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("X", "The input tensor of reshape operator.");
+    AddOutput("Out", "The output tensor of reshape operator.");
+    AddAttr<std::vector<int>>("shape", "Target shape of reshape operator.");
+    AddComment(R"DOC(Reshape operator
+
+Reshape Input(X) into the shape specified by Attr(shape).
+)DOC");
```
Comment (qingqing01) [inline paddle/operators/reshape_op.cc]: Need more comments : https://github.com/PaddlePaddle/Paddle/pull/3765/files#diff-6e4a3431c20e09cd11b1b56451b5754eR52

doc要求： https://github.com/PaddlePaddle/Paddle/issues/3885#issuecomment-327178855

File: paddle/operators/reshape_op.cc [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+class ReshapeGradOp : public framework::OperatorWithKernel {
+ public:
+  ReshapeGradOp(const std::string &type,
+                const framework::VariableNameMap &inputs,
+                const framework::VariableNameMap &outputs,
+                const framework::AttributeMap &attrs)
+      : OperatorWithKernel(type, inputs, outputs, attrs) {}
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
```
Comment (qingqing01) [inline paddle/operators/reshape_op.cc]: Need to check nonempty for the inputs.

Comment (kuke) [reply paddle/operators/reshape_op.cc]: Done

File: paddle/operators/reshape_op.cc (hunk 5 above)
Comment (kuke) [reply paddle/operators/reshape_op.cc]: Modified

File: paddle/operators/reshape_op.cc [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    auto shape = ctx.Attr<std::vector<int>>("shape");
+    int64_t capacity = -1;
+    for (auto dim : shape) {
+      PADDLE_ENFORCE(dim > 0, "Each dimension of shape must be positive.");
+      if (capacity < 0) {
+        capacity = dim;
+      } else {
+        capacity *= dim;
+      }
+    }
```
[the remark this answers is not in the document]
Comment (kuke) [reply paddle/operators/reshape_op.cc]: Done

File: paddle/operators/reshape_op.cc [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+class ReshapeOp : public framework::OperatorWithKernel {
+ public:
+  ReshapeOp(const std::string &type, const framework::VariableNameMap &inputs,
+            const framework::VariableNameMap &outputs,
+            const framework::AttributeMap &attrs)
+      : OperatorWithKernel(type, inputs, outputs, attrs) {}
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    auto *in = ctx.Input<framework::Tensor>("X");
```
[the remark this answers is not in the document]
Comment (kuke) [reply paddle/operators/reshape_op.cc]: Done

File: paddle/operators/reshape_op.cc (hunk 6 above)
Comment (kuke) [reply paddle/operators/reshape_op.cc]: Done

**Pull request merged**